Wait until the locator is ready before accepting client connections

kitcambridge opened the pull request:
This pull request adds an optional `ReadyNotifier` interface that locators can use to defer accepting client connections. Closes #111.

kitcambridge commented on src/github.com/mozilla-services/pushgo/simplepush/etcd_locator.go:
```diff
@@ -179,11 +194,26 @@ func (l *EtcdLocator) Init(app *Application, config interface{}) (err error) {
 
 	l.closeWait.Add(2)
+	if l.startDelay > 0 {
+		l.closeWait.Add(1)
```
This ensures `ReadyNotify()` always receives a value before the router is closed (i.e., the server is closed before it finishes starting up).

kitcambridge commented on src/github.com/mozilla-services/pushgo/simplepush/etcd_locator.go:
```diff
@@ -45,6 +45,11 @@ type EtcdLocatorConf struct {
+	// StartDelay is the amount of time to wait after registering the host with
+	// etcd. This should be 1-2 times the refresh interval to ensure the locator
+	// has a complete view of the cluster.
+	StartDelay string `toml:"start_delay" env:"start_delay"`
```
We could make this a multiplier, too; I just used an explicit duration to match `CloseDelay`.

bbangert commented:
Looks like it will probably work, but it'd be nice to have a test for a little more confidence. I have the sad thought that it will take more work to come up with a test.

kitcambridge commented:
> I have the sad thought that it will take more work to come up with a test.

Yes. :joy_cat: Some of the fake helpers for #179 might be useful here; I'll take a look in a bit.

kitcambridge commented:
Added tests, and fixed an issue where we were waiting for the full timeout if no one accepted a routed update. I think that should reduce some of the latency spikes we were seeing in the load tests.

This uses the mock helpers from #211; the commit set should look a lot cleaner once that's merged.

bbangert commented on src/github.com/mozilla-services/pushgo/simplepush/router_broadcast.go:
```diff
@@ -460,27 +440,28 @@ func (r *BroadcastRouter) notifyAll(cancelSignal <-chan bool, contacts []string,
+		go r.notifyContact(results, url, segment, logID)
+	}
+	timer := time.After(timeout)
+	for i := 0; i < cap(results) && !ok; i++ {
```
This actually represents whether it was delivered or not, so it should be 'delivered', not 'ok'.

bbangert commented:
r-, pending naming nit

bbangert commented:
r+
